Allow changing shade slot cap + prefer maintaining data

SupersonicNK opened the pull request:
Shade slots above the normal cap (8) are still readable/writable in rivals, but aren't used by the color shader. You can however use them to store miscellaneous data. Reading them is done using the functions `get_profile_color_slot_r/g/b`
For example, I use it to store the two endpoints of a color transition in an upcoming character.

I also modified the parsing functions to increase the maximum slot values rather than destroy the additional colors and shades, as it makes for less steps when importing a colors.gml file with more than the normal amount of data, because that additional data is still usable in-game.

additionally, a warning was added to color and shade slots above the game's technical maximum.

SupersonicNK committed:
change how shade rows are capped

effectively removes the cap; same method as changing the max palette number.
anything above 8 shade slots will not be matched, accurate to how the game functions. it will also display a little [!] to the left of that shade row to indicate that the color data will not be shown.
this data, like the additional palette data past the maximum is actually stored by the game and usable with the functions get_profile_color_slot_r/g/b, which is something i personally use in my characters but have not seen used very often.

SupersonicNK committed:
changed text import to maintain data + added cap warning

personal preference. left the old functionality in, but commented out.
also added warning indicator to color slots above the game's maximum.

CL-9a commented on index.js:
```diff
@@ -448,8 +450,11 @@ const vm = new Vue({
+
 			if (this.colorProfilesMainColors.length > this.MAX_ALT_PALETTES)
-				this.colorProfilesMainColors.splice(this.MAX_ALT_PALETTES, this.colorProfilesMainColors.length - 1)
+				this.MAX_ALT_PALETTES = this.colorProfilesMainColors.length; //prefer maintaining user data.
```
also not sure this comment's really useful in the context of a comment in the code - since it says "prefer" it refers to the change you've made in the commit, but a better place to do that would be to do it in the PR or in the commit message (and you did both already :muscle:)

(sorry, that's kinda nitpicking haha, but since I'm already asking to just clean up the commented code, I might as well!!)

CL-9a commented on index.js:
```diff
@@ -470,7 +475,10 @@ const vm = new Vue({
 
 			try {
-				this.rows = json.splice(0, this.MAX_SHADE_ROWS);
+				//this.rows = json.splice(0, this.MAX_SHADE_ROWS);
```
same here, can you remove old code?

CL-9a commented on index.js:
```diff
@@ -448,8 +450,11 @@ const vm = new Vue({
 			})
 
 			//only keep 16 color profiles
+			//if (this.colorProfilesMainColors.length > this.MAX_ALT_PALETTES)
+			//	this.colorProfilesMainColors.splice(this.MAX_ALT_PALETTES, this.colorProfilesMainColors.length - 1)
+
```
Could you just straight up remove the old code instead of commenting it? It's tracked in git anyway so it's not lost, and it makes it easier to see why stuff is here

CL-9a commented on index.js:
```diff
@@ -879,6 +889,54 @@ const vm = new Vue({
+									break; //prevent coloring anything past shade slot 7, as the game would not do.
+								const hsv = rgbToHsv(r, g, b);
+
+								//those ranges are precalculated in generateGmlCode so that we don't have to math them here
+								if(isWrappingValueWithinRange(hsv.h, 0, 360, rangeDef.hL, rangeDef.hH)
+								&& hsv.s >= rangeDef.sL && hsv.s <= rangeDef.sH
+								&& hsv.v >= rangeDef.vL && hsv.v <= rangeDef.vH
+								) {
+									const mainColorForShade = this.colorProfilesMainColors[this.selectedColorProfile].shades[shadeIndex];
+									matched = true;
+									const defaultColorForShade = this.colorProfilesMainColors[0].shades[shadeIndex];
+
+
+									const accurateHSV = rgbToHsv_noRounding(r, g, b);
+
+									const defaultToCurrentDeltaHSV = getHSVDelta(defaultColorForShade.accurateHSV, accurateHSV);
+									const shiftedHSV = applyDeltaToHSV(mainColorForShade.accurateHSV, defaultToCurrentDeltaHSV);
+
+									const shiftedRgb = hsvToRgb_noRounding(shiftedHSV.h, shiftedHSV.s, shiftedHSV.v);
+									const mainToShiftedDeltaRGB = getRGBDelta(mainColorForShade.rgb, shiftedRgb);
+									const shiftedRgb2 = applyDeltaToRGB({...shiftedRgb}, {
+										r: mainToShiftedDeltaRGB.r * (this.shadingValue - 1),
+										g: mainToShiftedDeltaRGB.g * (this.shadingValue - 1),
+										b: mainToShiftedDeltaRGB.b * (this.shadingValue - 1)
+									});
+
+
+									imageDataArray[i] = shiftedRgb2.r;
+									imageDataArray[i+1] = shiftedRgb2.g;
+									imageDataArray[i+2] = shiftedRgb2.b;
+
+									cachedColorTransforms.set(`${r},${g},${b}`, shiftedRgb2);
+
+									//console.log("px", i, "fitting rangeDef", hsv, mainColorForShade.hsv, defaultToCurrentDeltaHSV, shiftedRgb)
+								}
+								
+								shadeIndex++;
+							}
+							*/
+
```
This is just replacing the `forEach` with a `for ... of` so that you can do `break`, right? Could you also clean it up, and possibly make a separate commit / PR for it?

Ideally all that logic would be in a separate file and there'd be tests where we can easily see if anything broke + the time it took, but... this isn't an ideal world and I've been lazy

SupersonicNK replied:
That's correct, yeah. I want to test it more thoroughly at some point as I'm still not sure of the performance impact between the two. 
`for ... of` might perform better in the case where there are several extra shade slots, but I don't think it'd be an impact more than negligible still.

CL-9a reviewed (changes requested):
how do reviews work i'm testing things out
(overall it's good to me :ok_hand: just want to clean up the commented-out stuff to avoid cluttering the already-messy code, and since git tracks the changes, if we want to rollback we can literally roll back a commit, so there's not much point in keeping old code around)

SupersonicNK committed:
cleaned commented old code

old habits die hard

SupersonicNK committed:
cleaned hastily tested code

removes a commented out conversion of the range check from `forEach` to `for ... of`.
it was hastily tested, so whether it has any effect on performance is questionable.
separate commit as requested.

SupersonicNK commented:
I cleaned out the commented code in two steps (the old code first, then the hastily tested block in a separate commit).
